lib/maths/analytics/CBoostedTreeFactory.cc

```diff
@@ -1308,19 +1364,29 @@ CBoostedTreeFactory CBoostedTreeFactory::constructFromDefinition(
 
     CBoostedTreeFactory factory{constructFromParameters(numberThreads, std::move(loss))};
 
+    auto is = validInputStream(dataSearcher);
```

Comment:
Calling this only once instead of twice is skipping over this code (as well as breaching encapsulation by making assumptions about which implementation an interface points to):

https://github.com/elastic/ml-cpp/blob/e8a44b7c4fe79d1c5ea07c6114ce7c0a65a9164f/lib/api/CSingleStreamSearcher.cc#L41-L44

I think you should put this method back to how it was before.

The cause of the crash is probably different interpretation of the zero byte between the two JSON documents. We probably need to wrap up Boost.JSON in some way so that our wrapper treats zero bytes as document separators. Otherwise there will probably be several places that need changing in Java as well as in C++.